### PR TITLE
Customize _get_tensor_rank for model export in stage3

### PR DESCRIPTION
**Pull request by pengwa**
### Customize _get_tensor_rank for model export in stage3

Weight/Params sizes are all (0), so exporter logic depending on input shape will fail. 

This PR override `_get_tensor_rank` function by retrieving the shape for weight differently.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py [hunk 1]
```diff
@@ -359,3 +360,48 @@ def _update_python_op_input_related_attributes(
     node.attribute.remove(dtype_attr)
     node.attribute.append(helper.make_attribute("input_tensor_ranks", input_tensor_ranks))
     node.attribute.append(helper.make_attribute("input_tensor_types", input_tensor_dtypes))
+
+
+@contextmanager
+def stage3_export_context(enable: bool, graph_execution_manager):
+    if enable is False:
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py:367]:
```suggestion
    if not enable:
```

File: orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py [hunk 2]
```diff
@@ -359,3 +360,48 @@ def _update_python_op_input_related_attributes(
...
+        )
+
+        try:
+            from torch.onnx._internal import _beartype
+
+            @_beartype.beartype
+            def _get_tensor_rank(x) -> Optional[int]:
+                ### Adapted from https://github.com/pytorch/pytorch/blob/185515368bcd7d94ac06ab1634f22b747b03c6d9/torch/onnx/symbolic_helper.py#L561
+                # Retrieve the real rank for the stage3 weights, because stage3 weights are all (0).
+                import typing
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py:387]:
```suggestion
                from typing import cast as typing_cast
```

File: orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py [hunk 3]
```diff
@@ -359,3 +360,48 @@ def _update_python_op_input_related_attributes(
...
+
+                input_name = x.debugName()
+                if input_name in graph_execution_manager._zero_stage3_param_map:
+                    rank = len(graph_execution_manager._zero_stage3_param_map[input_name].ds_shape)
+                    return rank
+
+                if not _is_tensor(x) or x.type() is None:
+                    return None
+                x_type = x.type()
+                x_type = typing.cast(_C.TensorType, x_type)
```
Comment (AdamLouly) [inline orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py:400]:
```suggestion
                x_type = typing_cast(_C.TensorType, x_type)
```

**Review by AdamLouly — COMMENTED** [review]
Nit: some changes to make it more readable and remove that CodeQL warning.
otherwise LGTM.

File: orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_zero_stage3_compatibility.py]: Fixed, thank you!:)

**Review by AdamLouly — APPROVED** [review]

**Review by ajindal1 — APPROVED** [review]
LGTM

**Pull request merged**